### ISSUE TITLE
Spec

### ISSUE THREAD

**Issue by SylvainCorlay** [ipython/ipywidgets#895]
Two important areas where widgets would benefit from a formal specification

1. A json schema for the static widget serialized manager.

   This should be the simplest part since we effectively use this to power http://jupyter.org/widgets.html

   - It is likely to be complemented with a semver requirement for model modules.
   - At the moment, only the front-end can generate that json, and we should also have the backend be able to do it.

2. The message spec for display / update / custom messages and asynchronous state requests.

   This will also require a detailed description of the custom serialization mechanism.

**Comment by SylvainCorlay** [comment]
Ping @rgbkrk

**Comment by SylvainCorlay** [comment]
I would love to have feedback on the form the spec should have

1. First item

 - documentation + jsonschema for the format.

2. Second item

 - is there a standard or a good model I should use for inspiration for the spec?

**Comment by rgbkrk** [comment]
Excellent to see this!

> is there a standard or a good model I should use for inspiration for the spec?

On the slightly more informal route there's [the jupyter message spec](http://jupyter-client.readthedocs.io/en/latest/messaging.html#general-message-format). I personally think that the widgets should carve out a space on that spec that is more tightly defined than just the comms. It can use the comms -- the missing piece is what those messages are and how they get invoked (who does the `comm_open`, what it means, etc.)

If more formal, there are RFCs and W3C docs. It doesn't have to be that though. This can be refined and updated in the same way that the jupyter message protocol and notebook format get updated. They do both have JSON Schema as well.

The original spec for the widgets was in [IPEP 23](https://github.com/ipython/ipython/wiki/IPEP-23:-Backbone.js-Widgets). The current implementation of the widgets certainly surpasses that and should wholly replace it. The primary goal is to have something that can be versioned and worked with over time.

**Comment by SylvainCorlay** [comment]
Notes for self (on the plane to Frankfurt)

## Widget Manager State Schema

used e.g. in https://github.com/jupyter/jupyter.github.io/blob/master/assets/pythreejs-example.json

```js
{
    "$schema": "http://json-schema.org/draft-04/schema#",
    "description": "Jupyter Interactive Widget State JSON schema.",
    "type": "object",
    "additionalProperties": true,
    "additionalProperties" : {
        "type": "object",
        "properties": {
            "model_name": { "type": "string" },
            "model_module": { "type": "string" },  // AMD loadable module for model implementation
            "state": {
                "type": "object",
                 "additional_properties": true
            },
            "views": {
                "type": "array"
            }
        },
        "required": ["model_name", "model_module", "state"],
        "additional_properties": false
    }
}
```

 - The serialized widget manager is a set of key-value pairs where the key is the unique id of the model and the value is the serialized **model**.  The schema for a model state requires specifying a *model name*, a *model module* and a *state*.
 - Items in the model state correspond to a model attribute the javascript side. (keys = attribute name, value = serialized attribute value)
 - The schema for model attributes is completely up to the model. A mechanism to specify custom serializers and deserializers is provided to make sense of the values for state attributes. If none is provided, the default json deserialization is used.
- On the python side, state attributes are specified as traitlets attributes. Those synced with the front-end must have `sync=True` as metadata. The serializers and deserializers are specified via the `to_json` and  `from_json` metadata properties.
- On the JavaScript side, state attributes are specified as backbone model attributes. Custom serializers and deserializers are specified as class attributes of the model.
 - *Some* custom serializers / deserializer pairs are provided in the core widget library. For example, if `widget_serialization` is used, strings prefixed by `IPY_MODEL_` are replaced with references to other models.

**Comment by SylvainCorlay** [comment]
New version of the spec since #898 

{
    "$schema": "http://json-schema.org/draft-04/schema#",
    "description": "Jupyter Interactive Widget State JSON schema.",
    "type": "object",
    "additionalProperties": true,
    "additionalProperties" : {
        "type": "object",
        "properties": {
            "model_name": { "type": "string" },
            "model_module": { "type": "string" },  // AMD loadable module for model implementation
            "model_module_version": { "type": "string" },  // semver range for model module
            "state": {
                "type": "object",
                    "additional_properties": true
            },
            "views": {
                "type": "array"
            }
        },
        "required": ["model_name", "model_module", "state"],
        "additional_properties": false
    }
}

**Comment by SylvainCorlay** [comment]
Closing since we now have

- the  `jupyter-widgets-schema` npm package holding the spec for static widget state and view.
- corresponding documentation on embedding.

I am openning another issue on the protocol spec.

**Comment by SylvainCorlay** [comment]
@rgbkrk Regarding the specs.

1. I think that we are pretty much code-complete on the specification for the widget mime types and the serialization formats.

    - A format json schema for the two mime types (`application/vnd.jupyter.widget-state+json` and `application/vnd.jupyter.widget-view+json`) provided in a separate npm package (`jupyter-widgets-schema`).

    - New APIs to generate those serialized format on the python side (in addition to those that we had on the js side)

    - Two sphinx extensions make use of these mime types and APIs for rendering widgets in sphinx documentation
       - `jupyter_sphinx.embed_widgets` provides sphinx directives for running the code from a code snippet and use the python-side API for generating the right serialized format.
       ![sphinx-screencast](https://cloud.githubusercontent.com/assets/2397974/21186430/f6c72288-c214-11e6-9044-b82dead0aee2.gif)

      - `nbsphinx` now supports embedding widgets when converting notebooks to sphinx documentation which we now use for ipywidgets' documentation itself: http://ipywidgets.readthedocs.io/en/latest/examples/Widget%20List.html

    - The rendering of widgets on nbviewer (if the notebook was generated with the current master branch of widgets, to be released in ipywidgets 6.0)
        http://nbviewer.jupyter.org/github/ipython/ipywidgets/blob/master/docs/source/examples/Widget%20List.ipynb

2. Now we need to tackle a formal spec for the messaging protocol.

Any pointer to specifications that have the sort of format that we should adopt would be useful. I had put the tmpnb-backed widgets project on the backburner but it would be a good use case for iterating on this.
